common.py

```diff
@@ -0,0 +1,208 @@
+
+        return connection
+
+    def run_command(self, command, use_sudo=False):
+        with self._get_connection() as connection:
+            logger.debug('Running `%s` on %s', command, self.private_ip)
+            result = (connection.sudo(command, warn=True, hide='stderr')
+                      if use_sudo else
+                      connection.run(command, warn=True, hide='stderr'))
+            if result.failed:
+                raise ClusterInstallError(str(result.stderr.encode('utf-8')))
+
+    def put_file(self, local_path, remote_path):
+        if not isfile(local_path):
+            raise ClusterInstallError(
+                '{} is not a file'.format(local_path))
+
+        with self._get_connection() as connection:
+            logger.debug('Copying %s to %s on host %a',
+                         local_path, remote_path, self.private_ip)
+            connection.put(expanduser(local_path), remote_path)
+
+    def get_file(self, remote_path, local_path):
+        with self._get_connection() as connection:
+            logger.debug('Copying %s to %s from host %a',
+                         remote_path, local_path, self.private_ip)
+            connection.get(remote_path, expanduser(local_path))
+
+    def put_dir(self, local_dir_path, remote_dir_path, override=False):
+        """Copy a local directory to a remote host.
+
+        :param local_dir_path: An existing local directory path.
+        :param remote_dir_path: A directory path on the remote host. If the
+                                path doesn't exist, it will be created.
+        :param override: If True and the remote directory path exists, then
+                         it will be deleted.
+                         If False and the remote directory path exists, then
+                         the files from the local directory will be added to
+                         the remote one.
+                         If the remote directory path doesn't exist, it
+                         doesn't have any effect.
+        """
+        if not isdir(local_dir_path):
+            raise ClusterInstallError(
+                '{} is not a directory'.format(local_dir_path))
+
+        logger.debug('Copying %s to %s on host %a',
+                     local_dir_path, remote_dir_path, self.private_ip)
+        if override:
+            self.run_command('rm -rf {}'.format(remote_dir_path))
+        self.run_command('mkdir -p {}'. format(remote_dir_path))
+        for file_name in os.listdir(local_dir_path):
+            object_path = join(local_dir_path, file_name)
+            if isfile(object_path):
+                self.put_file(join(local_dir_path, file_name), remote_dir_path)
+            elif isdir(object_path):
+                self.put_dir(object_path, join(remote_dir_path, file_name))
+
+    def path_exists(self, path):
+        with self._get_connection() as connection:
+            sftp_client = connection.sftp()
+            try:
+                sftp_client.open(path)
```

Comment:
I recommend using sudo test -e <path>
sftp_client.open will not work with nested directories whose parents don't give you the necessary (execute) permissions, which may happen if you use your directory creation on a system with a restrictive umask.

Reply:
Done.